stingray/stats.py

```diff
@@ -776,3 +786,332 @@ def _pavnosigfun(power, nspec):
         m -= 1
 
     return sum
+
+
+def power_confidence_limits(preal, n=1, alpha=0.16):
+    """Confidence limits on power, given a signal power in the PDS/Z search.
+
+    Adapted from Vaughan et al. 1994, noting that, after appropriate
+    normalization of the spectral stats, the distribution of powers in the PDS
+    and the Z^2_n searches is always described by a noncentral chi squared
+    distribution.
+
+    Parameters
+    ----------
+    preal: float
+        The underlying, signal-generated value of power
+
+    Other Parameters
+    ----------------
+    n: int
+        The number of summed powers to obtain pmeas. It can be multiple
+        harmonics of the PDS, adjacent bins in a PDS summed to collect all the
+        power in a QPO, or the n in Z^2_n
+    alpha: float
+        The p-value (e.g. 0.16 = 16% for 1-sigma)
+
+    Results
+    -------
+    pmeas: [float, float]
+        The upper and lower confidence interval (a, 1-a) on the measured power
+
+    Examples
+    --------
+    >>> cl = power_confidence_limits(150)
+    >>> np.allclose(cl, [127, 176], atol=1)
+    True
+    """
+    rv = stats.ncx2(2 * n, preal)
+    return rv.ppf([alpha, 1 - alpha])
+
+
+def power_upper_limit(pmeas, n=1, c=0.95):
+    """Upper limit on signal power, given a measured power in the PDS/Z search.
+
+    Adapted from Vaughan et al. 1994, noting that, after appropriate
+    normalization of the spectral stats, the distribution of powers in the PDS
+    and the Z^2_n searches is always described by a noncentral chi squared
+    distribution.
+
+    Note that Vaughan+94 gives p(pmeas | preal), while we are interested in
+    p(real | pmeas), which is not described by the NCX2 stat. Rather than
+    integrating the CDF of this probability distribution, we start from a
+    reasonable approximation and fit to find the preal that gives pmeas as
+    a (e.g.95%) confidence limit.
+
+    As Vaughan+94 shows, this power is always larger than the observed one.
+    This is because we are looking for the maximum signal power that,
+    combined with noise powers, would give the observed power. This involves
+    the possibility that noise powers partially cancel out some signal power.
+
+    Parameters
+    ----------
+    pmeas: float
```

Comment:
So this is a bit confusing: in the previous function, `pmeas` describes the confidence intervals, here it describes the input measured power, for which I think you used `preal` in the previous function? I think it'd be helpful for both us and the user to unify the notation and make sure we use the same variables for the same quantities wherever possible.

Reply:
In the previous function, we wanted an expected range of measured powers given a theoretical value (e.g. from a modulation measured in previous observations). Here, we have a measured power and we want the signal power that could produce the measured power (or above) to a given confidence level